lib/Scheduling/Scheduler.cpp

```diff
@@ -0,0 +1,156 @@
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkDep(Dependence *dep) {
+  Operation *i = dep->getSource();
+  Operation *j = dep->getDestination();
+  if (!(hasOperation(i) && hasOperation(j))) {
+    containingOp->emitError()
+        << "Scheduling problem contains dependence with unregistered endpoints."
+        << "\n  from: " << *i << (hasOperation(i) ? "" : " (unregistered)")
+        << "\n  to:   " << *j << (hasOperation(j) ? "" : " (unregistered)");
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkOpr(OperatorType *opr) {
+  if (!hasLatency(opr)) {
+    containingOp->emitError()
+        << "Operator type '" << opr->getName() << "' has no latency";
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkProb() { return success(); }
+
+/// Check overall problem by delegating to the component-specific checkers.
+LogicalResult Scheduler::check() {
+  for (auto *op : getOperations()) {
+    if (failed(checkOp(op)))
+      return failure();
+  }
+
+  for (auto *dep : getDependences()) {
+    if (failed(checkDep(dep)))
+      return failure();
+  }
+
+  for (auto *opr : getOperatorTypes()) {
+    if (failed(checkOpr(opr)))
+      return failure();
+  }
+
+  return checkProb();
+}
+
+LogicalResult Scheduler::verifyOp(Operation *op) {
+  if (!hasStartTime(op)) {
+    op->emitError("Operation has no start time");
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::verifyDep(Dependence *dep) {
+  Operation *i = dep->getSource();
+  Operation *j = dep->getDestination();
+
+  unsigned stI, latI, stJ;
+  stI = getStartTime(i);
+  latI = getLatency(getAssociatedOperatorType(i));
+  stJ = getStartTime(j);
+
+  // check if i's result is available before j starts
+  if (!(stI + latI <= stJ)) {
+    containingOp->emitError()
+        << "Precedence violated for dependence."
+        << "\n  from: " << *i << ", result available in t=" << (stI + latI)
+        << "\n  to:   " << *j << ", starts in t=" << stJ;
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::verifyOpr(OperatorType *opr) { return success(); }
+
+LogicalResult Scheduler::verifyProb() { return success(); }
+
+/// Verify overall solution by delegating to the component-specific verifiers.
+LogicalResult Scheduler::verify() {
+  for (auto *op : getOperations()) {
+    if (failed(verifyOp(op)))
+      return failure();
+  }
+
+  for (auto *dep : getDependences()) {
+    if (failed(verifyDep(dep)))
+      return failure();
+  }
+
+  for (auto *opr : getOperatorTypes()) {
+    if (failed(verifyOpr(opr)))
+      return failure();
+  }
+
+  return verifyProb();
+}
+
+Dependence *
+Dependence::construct(mlir::StorageUniquer::StorageAllocator &allocator,
+                      const KeyTy &key) {
+  auto *result = allocator.allocate<Dependence>();
+  std::tie(result->src, result->dst, result->srcIdx, result->dstIdx) = key;
```

Comment:
I don't know how I feel about `std::tie`. I usually avoid it, but this is a case where I actually think it might be worth it :thinking:

Reply:
Saves 3 LoC here 😉 I'll revisit this after the discussion of the representation of dependences is resolved.

Reply:
My need for the `tie` went away 🤓